js/app/dashboard/redux/reducers.js

```diff
@@ -123,50 +125,53 @@ export const polylineEnabled = (state = {}, action) => {
 }
 
 export const selectedTasks = (state = [], action) => {
+  /*
+    FIXME selectedTasks is an array of task ids, not tasks objects
+  */
+  let newSelectedTasks
+
   switch (action.type) {
-  case TOGGLE_TASK:
+    case TOGGLE_TASK:
 
-    if (-1 !== state.indexOf(action.task['@id'])) {
-      if (!action.multiple) {
-        return []
+      if (-1 !== state.indexOf(action.task['@id'])) {
+        // let's remove this task!
+        if (!action.multiple) {
+          return []
+        }
+        return _.filter(state, task => task !== action.task['@id'])
       }
-      return _.filter(state, task => task !== action.task['@id'])
-    }
-
-    const newState = action.multiple ? state.slice(0) : []
-    newState.push(action.task['@id'])
-
-    return newState
 
+      newSelectedTasks = action.multiple ? action.currentlySelectedTasks.slice(0) : []
+      newSelectedTasks.push(action.task)
+      break
   case SELECT_TASK:
-
     if (-1 !== state.indexOf(action.task['@id'])) {
-
       return state
     }
 
-    return [ action.task['@id'] ]
-
+    return [action.task['@id']]
   case SELECT_TASKS:
-
-    return action.tasks.map(task => task['@id'])
-
-  case SELECT_TASKS_BY_IDS:
-
-    return action.taskIds
-
+    newSelectedTasks = action.tasks
+    break
   case CLEAR_SELECTED_TASKS:
   case MODIFY_TASK_LIST_REQUEST_SUCCESS:
-
     // OPTIMIZATION
     // Make sure the array if not already empty
     // before returning a new reference
     if (state.length > 0) {
       return []
     }
-    break
+  break
   }
 
+  if (newSelectedTasks && newSelectedTasks.length > 0) {
+    if (isValidTasksMultiSelect(newSelectedTasks)) {
+      return newSelectedTasks.map(task => task['@id'])
+    } else {
+      toast.warn(i18next.t('ADMIN_DASHBOARD_INVALID_TASKS_SELECTION'), {autoclose: 15000})
```

Comment:
I would avoid touching UI from a reducer, let's move it at least to the action

Reply:
I put it there because they are 3 different actions touching this slice of state, so I wanted to have this on one place only.

and it is pretty capital that no one breaks this unintentionnaly because i rely on it in other places of the code.

what do you think?

Reply:
I don't know if it's the best approach, but as we have only ids in the state and you need task objects to perform the validation. I would put all of these inside action creators and make it a function if you need it for multiple actions

Reply:
I think it will be more clean rather than adding state into the actions:
```
const currentlySelectedTasks = selectSelectedTasks(getState())
dispatch({ type: TOGGLE_TASK, task, multiple, currentlySelectedTasks })
```

Reply:
I will do in my next PR